packages/compass-editor/src/json-editor.tsx

```diff
@@ -838,11 +900,45 @@ const BaseEditor = React.forwardRef<EditorRef, EditorProps>(function BaseEditor(
         readOnlyExtension,
         EditorView.updateListener.of((update) => {
           updateEditorContentHeight();
+          const editorText = editor.state.sliceDoc() ?? '';
+
+          if (update.focusChanged && editorViewRef.current?.hasFocus) {
+            const editorView = editorViewRef.current;
+
+            onFocusRef.current?.({ editorView, setContent }, editorText);
```

Comment:
SetContent is a wrapper on CodeMirror dispatch, to avoid coupling more components to CodeMirror. We could use `useImperativeHandle`, never used it, but what is the benefit of it? Is adding setContent to useLayoutEffect enough to ensure that there is no particular side-effect we shouldn't expect?

Reply:
1. You don't need to use redux state in the callbacks provided by parents at all, can you explain why are you keep bringing it up?
2. The `useImperativeHandle` hook is exactly the pattern in React to [provide custom imperative interfaces to the parent components](https://react.dev/reference/react/useImperativeHandle), instead what you are doing is baking in a lot of custom logic inside a component that doesn't need to handle it and exposing it in a non-standard way. This hook is already used in this component and we should not break the pattern, I see no reason to. 
3. Not that relevant because we need to change this, but yes, adding `setContent` to the dependencies would be a right move here, this effect is responsible for mounting the editor though, so it should not ever run more than once, as described in the code there. A more straightforward way of handling it here, as editor instance is the only dependency of this function, would be to just move the definition inside the effect